### PR TITLE
OSDEV-3326: [CI/CD] Fix deployment pipeline

### PR DESCRIPTION
**Pull request by valencia-max**
## Jira Ticket

N/A

---

## Summary of Changes

The `kafka` bake target has been failing the deploy pipeline at the extract step:

```
ERROR: target kafka: failed to solve: process "/bin/sh -c tar -xzf kafka_2.13-3.8.0.tgz"
did not complete successfully: exit code: 2
```

**Root cause: an unpinned base image moved underneath us.** `src/kafka-tools/Dockerfile` used
`FROM amazonlinux:latest`, and that tag has moved from Amazon Linux 2 to AL2023. AL2 shipped
`gzip` in the base image; AL2023 does not. `tar -xz` shells out to the `gzip` binary, so without
it tar fails with `gzip: Cannot exec` → `Error is not recoverable: exiting now` → **exit code 2**.
Line 4 installed `tar` but never `gzip`.

The download on line 5 was never the problem. It also explains why the failure looked like it
appeared out of nowhere with no change to the file: `java-1.8.0` still resolves on AL2023 (the
Corretto 8 RPM carries that virtual provide), so the `yum install` line kept passing and the
breakage surfaced one line later.

Changes, in order of importance:

**The fix**

- Pin the base to `amazonlinux:2023` instead of `:latest`, so the image cannot change majors
  again without a commit.
- Install `gzip` explicitly, plus `gawk` — `fix_replication.sh` parses `--describe` output with
  awk functions and that dependency was implicit.
- Name the JRE package exactly: `java-1.8.0-amazon-corretto`. Same package the bare `java-1.8.0`
  was already resolving to; there is no `-headless` variant for Corretto 8 on AL2023.

**Fail-loud downloading** (so the next transport problem reports itself instead of surfacing as a
confusing `tar` error)

- `curl -f` replaces `wget`. `wget` exits 0 when a server answers a rate-limited or bot-checked
  request with an HTML body under HTTP 200 — it writes that HTML to `kafka_*.tgz` and the build
  only fails later at `tar`. `-f` plus a SHA-512 comparison against Apache's published `.sha512`
  makes the download step itself fail, and print the first 200 bytes of whatever actually arrived.
- `--proto '=https' --proto-redir '=https'` on both `curl` calls. `KAFKA_MIRROR` is a build arg,
  so an `http://` override was reachable, and `-L` would otherwise follow an HTTPS→HTTP
  downgrade redirect. (Raised in review.)
- Kafka version, Scala version, and mirror are now `ARG`s rather than four hard-coded literals.

**Run unprivileged** (raised in review — the image ran as root, and `task-definitions/app_kafka.json`
sets no `user`, so `./kafka.sh` inherited root in every deploy)

- Add a system user `kafka` (uid/gid 10001) and `USER kafka`.
- `chown` is applied to the `${KAFKA_HOME}` **directory only, deliberately not `-R`**. The extracted
  distribution is already world-readable (`bin/*.sh` 755, libs 644), so recursion would duplicate
  ~100 MB into a new layer purely to rewrite ownership bits. Write permission on the directory is
  the part actually needed, because `fix_replication.sh` writes `add-replicas.json` into its CWD.
- `ENV LOG_DIR=/tmp/kafka-logs`. Every `kafka-*.sh` wrapper sources `kafka-run-class.sh`, which
  runs `mkdir -p "$LOG_DIR"` with `LOG_DIR` defaulting to `$KAFKA_HOME/logs`. Redirecting it keeps
  that write off the distribution directory rather than relying on the chown to absorb it.
- `COPY --chown=kafka:kafka` for the four script/config files.
---

## Type of Change

- [x] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Backend change (API, database, business logic)
- [ ] UI / frontend change
- [ ] Architectural change (affects structure, contracts, or shared modules)
- [ ] Refactor (no functional change)
- [ ] Breaking change (existing functionality may not work as before)
- [x] Documentation / tooling only

---

## Testing
Deploy to AWS was succesfully: https://github.com/opensupplyhub/open-supply-hub/actions/runs/32692175440

---

## Known TODO Items

NA

---

### Implementation

- [x] My code follows the project's style guidelines and naming conventions.
- [x] I have removed debug logs, commented-out code, and unused imports.
- [x] Any AI-assisted code (e.g., Claude) has been reviewed, understood, and adapted to fit our codebase.
- [x] I have considered backward compatibility and migrations where applicable.
- [x] The diff contains no secrets, internal document/spreadsheet/folder IDs, internal board or channel identifiers, internal policy thresholds, or staff contact details — internal values are externalized to runtime configuration (see "Public repository hygiene" in AGENTS.md).

### Testing & Validation

- [x] I have manually tested the change in a local/dev environment.
- [x] All existing tests pass locally.
- [ ] For backend changes: I have validated API contracts, error handling, and edge cases.

### Documentation & Communication

- [x] I have updated relevant documentation (README, ADRs, inline comments, API docs).
- [x] I have updated the Jira ticket status and added any relevant notes.

### Final Review

- [x] I have performed a self-review on my PR and I am presenting my best work for my peers to review.

### CODE REVIEW

File: src/kafka-tools/Dockerfile [hunk 1]
```diff
@@ -1,11 +1,70 @@
...
+
+# curl -f rather than wget: archive.apache.org can answer a rate-limited or
+# bot-checked request with an HTML page under HTTP 200. wget saves that as
+# kafka_*.tgz and exits 0, so the build only blows up later at `tar`, with no
+# hint of why. -f plus the checksum comparison make the download step itself
+# fail, and print what actually arrived.
+RUN set -eu; \
+    tarball="kafka_${KAFKA_SCALA_VERSION}-${KAFKA_VERSION}.tgz"; \
+    url="${KAFKA_MIRROR}/${KAFKA_VERSION}/${tarball}"; \
+    curl -fsSL --retry 5 --retry-delay 5 --retry-connrefused -o "${tarball}" "${url}"; \
```
[the remark this answers is not in the document]
Comment (valencia-max) [reply src/kafka-tools/Dockerfile]: Done

File: src/kafka-tools/Dockerfile [hunk 2]
```diff
@@ -1,11 +1,70 @@
...
+    expected="$(curl -fsSL --retry 5 --retry-delay 5 "${url}.sha512" \
```
[the remark this answers is not in the document]
Comment (valencia-max) [reply src/kafka-tools/Dockerfile]: Done

File: src/kafka-tools/Dockerfile [hunk 3]
```diff
@@ -1,11 +1,70 @@
-FROM amazonlinux:latest
+# Kafka CLI tools image: provides kafka-topics.sh / kafka-configs.sh /
+# kafka-reassign-partitions.sh for kafka.sh (topic creation) and
+# fix_replication.sh (OSDEV-3063).
+
+# Pinned, not :latest. `amazonlinux:latest` moves between major versions, and
+# AL2 -> AL2023 changes both the package set and what ships in the base image
+# (notably: no gzip). A moving base is why a build that passed last month can
+# fail with no change to this file.
+FROM amazonlinux:2023
```
[the remark this answers is not in the document]
Comment (valencia-max) [reply src/kafka-tools/Dockerfile:9]: Done

File: src/kafka-tools/Dockerfile [hunk 4]
```diff
@@ -1,11 +1,70 @@
...
+# bot-checked request with an HTML page under HTTP 200. wget saves that as
+# kafka_*.tgz and exits 0, so the build only blows up later at `tar`, with no
+# hint of why. -f plus the checksum comparison make the download step itself
+# fail, and print what actually arrived.
+RUN set -eu; \
+    tarball="kafka_${KAFKA_SCALA_VERSION}-${KAFKA_VERSION}.tgz"; \
+    url="${KAFKA_MIRROR}/${KAFKA_VERSION}/${tarball}"; \
+    curl -fsSL --retry 5 --retry-delay 5 --retry-connrefused -o "${tarball}" "${url}"; \
+    expected="$(curl -fsSL --retry 5 --retry-delay 5 "${url}.sha512" \
+        | tr -d '[:space:]' | tr 'A-F' 'a-f' | sed 's/^.*tgz://')"; \
```
[the remark this answers is not in the document]
Comment (valencia-max) [reply src/kafka-tools/Dockerfile]: Done

File: src/kafka-tools/Dockerfile [hunk 5]
```diff
@@ -1,11 +1,70 @@
...
+        exit 1; \
+    fi; \
+    tar -xzf "${tarball}"; \
+    rm -f "${tarball}"
+
+# Symlink, not a hard link: a hard link silently breaks if the target layer is
+# rebuilt, and /usr/local/bin is the conventional place for this.
+RUN ln -s "${KAFKA_HOME}/bin/kafka-topics.sh" /usr/local/bin/kafka-topics
+
+WORKDIR ${KAFKA_HOME}
```
[the remark this answers is not in the document]
Comment (valencia-max) [reply src/kafka-tools/Dockerfile:93]: Done

**Review by protsack-stephan — APPROVED** [review]
